compiler/rustc_resolve/src/imports.rs

```diff
@@ -955,14 +955,18 @@ impl<'a, 'b> ImportResolver<'a, 'b> {
                 }
                 return None;
             }
-            PathResult::NonModule(path_res) if path_res.base_res() == Res::Err => {
-                if no_ambiguity {
-                    assert!(import.imported_module.get().is_none());
+            PathResult::NonModule(path_res) => {
+                if path_res.base_res() == Res::Err {
+                    if no_ambiguity {
+                        assert!(import.imported_module.get().is_none());
+                    }
+                    // The error was already reported earlier.
+                    return None;
+                } else {
+                    return None;
                 }
-                // The error was already reported earlier.
-                return None;
             }
```

Comment:
I think it should be
```rust
            PathResult::NonModule(path_res) => {
                if no_ambiguity {
                    assert!(import.imported_module.get().is_none());
                }
                // The error was already reported earlier.
                return None;
            }
```
`Res::Err` was used to discern between reachable cases and (seemingly) unreachable cases.

Reply:
By the way, why do `clippy::a` and `clippy::a::b` seem to go through very different code paths? `clippy::a` didn't trigger the ICE, only `clippy::a::b`.

Reply:
Not sure what exactly happens there, but both imports and tool attributes are "special" in some regards.
- Import paths are split into two parts - module (everything except for the last segment, type namespace) and the final segment resolved multiple times in all namespaces.
- Tool attributes are also split into two parts but in a different way - the first segment resolved as a tool module, and all other segments that are not even resolved (because no definitions for them exist in source code), just automatically assumed to be a tool attribute. (Perhaps the implementation is taking a shortcut here and it could be done better, at least for diagnostics.)

When you combine these two special cases together you get some unique code paths.